readium/navigator/src/main/java/org/readium/r2/navigator/media2/SessionPlayerFacade.kt

```diff
@@ -0,0 +1,200 @@
+package org.readium.r2.navigator.media2
+
+import android.app.PendingIntent
+import android.content.Context
+import androidx.media2.common.MediaItem
+import androidx.media2.common.MediaMetadata
+import androidx.media2.common.SessionPlayer
+import androidx.media2.session.MediaSession
+import org.readium.r2.navigator.ExperimentalAudiobook
+import org.readium.r2.shared.util.Try
+import org.readium.r2.shared.util.flatMap
+import timber.log.Timber
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.ExperimentalTime
+
+/**
+ * This class's purpose is two-fold:
+ * - wrapping the [SessionPlayer] inside a coroutine-based API
+ * - adding compound commands
+ *
+ * All commands, either basic or compound, are executed in order of arrival on the same thread.
+ * Compound commands have two limitations due to the framework weaknesses:
+ * - in case of failure, the playback can be left in an intermediate state
+ * - the behaviour is undefined if any external controller takes actions at the same time
+ */
+@ExperimentalAudiobook
+@OptIn(ExperimentalTime::class)
+internal class SessionPlayerFacade(
+    private val sessionPlayer: SessionPlayer,
+) {
+    val playerState: SessionPlayerState
+        get() = SessionPlayerState.from(sessionPlayer.playerState)
+
+    val currentPosition: Duration?
+        get() = sessionPlayer.currentPositionDuration
+
+    val currentIndex: Int?
+        get() = sessionPlayer.currentMediaItemIndex
+            .takeUnless { it == SessionPlayer.INVALID_ITEM_INDEX }
+
+    val currentDuration: Duration?
+        get() = sessionPlayer.duration
+            .takeUnless { it == SessionPlayer.UNKNOWN_TIME }
+            ?.milliseconds
+
+    val playlist: List<MediaItem>?
+        get() = sessionPlayer.playlist
+
+    val playbackSpeed: Double?
+        get() = sessionPlayer.playbackSpeedNullable
+
+    fun unregisterPlayerCallback(callback: SessionPlayer.PlayerCallback) {
+        sessionPlayer.unregisterPlayerCallback(callback)
+    }
+
+    fun close() {
+        sessionPlayer.close()
+    }
+
+    fun session(context: Context, id: String, activityIntent: PendingIntent): MediaSession {
+        return MediaSession.Builder(context, sessionPlayer)
+            .setId(id)
```

Comment:
According to the doc:

> Use this **if and only if** your app supports multiple playback at the same time and also wants to provide external apps to have finer controls of them.
> https://developer.android.com/reference/androidx/media2/session/MediaSession.Builder#setId(java.lang.String)

So maybe we should use `id: String? = null` and set it only if non-null.

Reply:
I can't see the documentation's point against IDs. I used the ID before, now not any more so I can remove the ID session.

Reply:
Done.